flipperdevices/flipperzero-firmware
# Badusb keyboardlayout support

Reopen #1525

# What's new

- added a config menu to the badusb application which allows to select the keyboard layout used when emulating a hid and running a badusb script. the selected layout is persisted and reused on subsequent script execution until another layout is selected. some keyboard layouts are provided, layout files can be generated using the tool ([here](https://github.com/dummy-decoy/flipperzero_badusb_kl))
- reworked directory structure on the sd card to avoid mixing scripts and keyboard layout files. new structure and additional sd card files are in "assets/resources/badusb".

# Verification 

- copy the content of the "assets/resources/badusb/" folder into the "badusb/" folder of the sd card. if the sd card does not have the new directory hierarchy, the display may be confusing because the file selector will go straight to the root of the sd-card (but still everything will work)
- on your computer, select a keyboard input language other than english-us.
- plug the flipperzero to a computer with an usb cable
- launch the badusb application, select a simple script
- press run. the script executes, if your computer keyboard is not set to "en-US", you may see some strange character replacement ('q' instead of 'a', 'z' instead of 'y'...)
- on the flipper, still on the script execution page, press the left button to enter the "config" menu. select a keyboard layout matching the one on your computer. after selecting a layout, you are back to the config menu, you should press back once again to see the script execution view again.
-  on the flipper, press "run". the script executes, you should see exactly what was intended, without character replacements.
- when you exit the badusb application, you should see a ".badusb.settings" file appear next to the script and layouts folders. this file persists the last used keyboard layout. application should work even if this file does not exist.


# Checklist (For Reviewer)
